javarosa-dag.rst

```diff
@@ -0,0 +1,304 @@
+
+    // Serialization methods
+
+    getConditions();
+    getRecalculates();
+
+    // Misc
+
+    getEvalBehavior();
+    getTriggerableForRepeatGroup(TreeReference ref);
+    printTriggerables(String path);
+
+    // Seems to be used only by a FormDef method which isn't used in JavaRosa
+
+    getConditionExpressionForTrueAction(FormInstance mainInstance, TreeElement instanceNode, int action);
+
+Initialization
+-----------------
+
+Collecting
+~~~~~~~~~~
+
+Triggerables are constructed during the `bind <https://opendatakit.github.io/xforms-spec/#bindings>`_ parsing.
+The ``StandardBindAttributesProcessor`` class constructs appropriate ``Triggerable`` implementations based on the bind 
+attributes and adds them to the DAG. The bind processor class calls ``FormDef.addTriggerable(Triggerable t)`` 
+which calls ``IDag.addTriggerable(Triggerable t)``. 
+The DAG code is intentionally hidden from the parsing module and communication is made through the ``FormDef`` class.
+
+Two or more nodes can have the same triggerable expression for a given bind attribute. For example:
+
+.. code-block:: xml
+
+    <instance>
+        <data id="dataid">
+            <nodeA/>
+            <nodeB/>
+            <nodeC/>
+        </data>
+    </instance>
+    <bind nodeset="/data/nodeA" calculate="pow(/data/nodeC, 2)" type="int"/>
+    <bind nodeset="/data/nodeB" calculate="pow(/data/nodeC, 2)" type="int" />
+    <bind nodeset="/data/nodeC" type="int"/>
+
+In the above snippet, both ``nodeA`` and ``nodeB`` have the same ``calculation`` expression which is evaluated
+when ``nodeC`` is updated. The ``nodeC`` is the *trigger* of ``nodeA`` and ``nodeB`` calculations.
+
+To avoid storing two separate but equal``Triggerable`` instances, the ``IDag.addTriggerable(Triggerable t)``
+method first checks if a same triggerable has been already added. 
+Two ``Triggerable`` instances are considered equal if:
+
+* Both have the same implementation type - ``Recalculate`` never equals ``Condition``.
+* Both have the same expressions to be evaluated.
+* Both have the same triggers. This is implied by the second bullet as triggers are the nodes referenced in the expressions.
+
+If such triggerable already exists, ``IDag`` does not add a new instance but updates the existing``Triggerable`` one's context so it points to the 
+highest common root. For the above example, the highest common root for ``nodeA`` and ``nodeB`` calculations is ``/data``.
+The passed instance ``t`` is ignored and a previously added ``Triggerable`` one is returned and should be used for the further processing.
+The context is used later when constructing a DAG structure.
+
+If such triggerable does not exist, it is added to the ``unorderedTriggerables`` list and the ``triggerIndex`` map.
+The passed triggerable ``t`` instance is returned.
+The list is later used to construct the topoligically sorted list ``triggerablesDAG``.
```

Comment:
> topoligically

I won’t flag any more spelling errors that should be caught by a spelling checker.

Reply:
@dcbriccetti Yeah that's right. I should've used a spelling checker. Will fix all of them. Thanks for all the errors you've caught so far!

Reply:
@dcbriccetti I think I fixed all simple spelling errors and apologize I hadn't done it earlier.
I also fixed other errors like missing articles or wrong words (e.g. infinite vs infinity) you've caught but there may be more typos like them.